### ISSUE TITLE
nomad CLI w/ TLS gives x509 error unless NOMAD_TLS_SERVER_NAME is set

### ISSUE THREAD

**Issue by avirshup** [hashicorp/nomad#7529]
### Nomad version
Nomad v0.10.4

### Operating system and Environment details
CLI running from OS X 10.14.6
Nomad server running on AWS in Amazon Linux 2.

### Issue
Apologies if I'm misunderstanding the expected behavior here.

I'm working on enabling TLS for our nomad cluster, and am directly following [the tutorial](https://learn.hashicorp.com/nomad/transport-security/enable-tls). [edited to add:] As instructed in the documentation, I issued the certificates by role and region, (i.e., not using FQDNs or IP addresses).

However, I get an x509 error when trying to run the examples in the tutorial:
```bash
> export NOMAD_ADDR=https://ec2-X-Y-Z.us-west-2.compute.amazonaws.com:4646
> export NOMAD_CLIENT_KEY=certs/cli-key.pem
> export NOMAD_CLIENT_CERT=certs/cli.pem
> export NOMAD_CACERT=certs/ca.pem
> nomad node status
Error querying node status: Get https://ec2-X-Y-Z.us-west-2.compute.amazonaws.com:4646/v1/nodes: x509: certificate is valid for server.global.nomad, localhost, not ec2-X-Y-Z.us-west-2.compute.amazonaws.com
```

However, everything works fine if I _also_ set NOMAD_TLS_SERVER_NAME and do it again:
```
> export NOMAD_TLS_SERVER_NAME=server.global.nomad
> nomad node status
> echo $?
0
```

### Reproduction steps
The certificates were generated using the exact commands from the tutorial at https://learn.hashicorp.com/nomad/transport-security/enable-tls. The sever's certificate is for `server.global.nomad`, `localhost`, and `127.0.0.1`.

Server configuration:
```hcl
name = "hobo-mark3-server"
datacenter = "hobo-mark3"
region = "global"
data_dir = "/data"
tls {
  http = true
  rpc  = true
  verify_https_client = true
  verify_server_hostname = true
  ca_file = "ca.pem"
  cert_file = "cert.pem"
  key_file = "cert-key.pem"
}
server {
  bootstrap_expect=1
  enabled=true
}
```

**Comment by greut** [comment]
It looks like you should add the fqdn in the certificate you're generating.

```
-hostname="ec2-X-Y-Z.us-west-2.compute.amazonaws.com,server.global.nomad,localhost,127.0.0.1"
```

TLS is hard ;)

**Comment by avirshup** [comment]
@greut wrote:
> It looks like you should add the fqdn in the certificate you're generating.
> TLS is hard ;)

No, the [Nomad documentation](https://learn.hashicorp.com/nomad/transport-security/enable-tls#node-certificates) explicitly says that you don't need to add the hostname or IP address to the certificate:

> TLS certificates commonly use the fully-qualified domain name of the system being identified as the certificate's Common Name (CN). However, hosts (and therefore hostnames and IPs) are often ephemeral in Nomad clusters. Not only would signing a new certificate per Nomad node be difficult, but using a hostname provides no security or functional benefits to Nomad. To fulfill the desired security properties (above) Nomad certificates are signed with their region and role such as:
>`   client.global.nomad for a client node in the global region`
>`   server.us-west.nomad for a server node in the us-west region`

**Comment by greut** [comment]
@avirshup In my understanding, the TLS connection is established before anything else can be done. Oo be able to use the `.nomad` name the resolution for that name must work, so the right name is sent (SNI and stuff). Make it work via, `NOMAD_ADDR=server.global.nomad`. To be honest with you, I didn't set up the `verify_server_hostname` settings yet, but you should be able to connect using the bare `openssl s_client`, as the nomad client cannot do more magic than that.

Best of luck!

**Comment by avirshup** [comment]
@greut - I'll repeat myself a third time - this is not what the documentation says, which is why I opened an issue.

**Comment by greut** [comment]
@avirshup the values goes right into Go TLS Config, https://pkg.go.dev/crypto/tls?tab=doc#Config (see ServerName). It makes sense that the names are private ones.

```console
$ echo | openssl s_client \
    -connect ec2-X-Y-Z.us-west-2.compute.amazonaws.com:4646

$ echo | openssl s_client \
    -connect ec2-X-Y-Z.us-west-2.compute.amazonaws.com:4646 \
    -servername server.global.nomad
```

My best guess is that the documentation doesn't tell the whole story. I don't see how it could be a bug as it has very few to do with Nomad itself. Cheers,

**Comment by mocofound** [comment]
@avirshup Please try using ```https://localhost:4646``` as your ```NOMAD_ADDR``` as indicated in the Learn guide.  For simplicity, the guide is written in a way that assumes you are directly connected to your Nomad cluster, for example SSH-ed into your Nomad server ec2 instance.  This way, you are trusted and don't have to worry about external SANs.

```export NOMAD_ADDR=https://localhost:4646```

As @greut mentioned, adding the ec2 instance external hostname looks like another way to do this if you wanted to use ```https://ec2-X-Y-Z.us-west-2.compute.amazonaws.com:4646``` for example to connect from _outside_ the cluster. This is covered in the Learn guide, although it could maybe be worded a little bit better.

> Using localhost and 127.0.0.1 as subject alternate names (SANs) allows tools like curl to be able to communicate with Nomad's HTTP API when run on the same host. Other SANs may be added including a DNS resolvable hostname to allow remote HTTP requests from third party tools.

In summary, I do think this is just a misunderstanding as it relates to intra-cluster vs extra-cluster communication.  Please let us know how it goes connecting directly to the cluster and using ```https://localhost:4646```!

**Comment by avirshup** [comment]
@mocofound - ah, gotcha, I did not understand that there's a difference between intra- and extra-cluster comms.

Setting `export NOMAD_ADDR=https://localhost:4646` doesn't work in my case, because I'm using the CLI on my local machine to connect to a server running on AWS. But since servers are going to be less ephemeral than clients, it does make sense to either set the `NOMAD_TLS_SERVER_NAME` or just go through the trouble of adding the FQDN to the certificates we generate for our servers. I appreciate your help!
